### PR TITLE
[libSyntax, tests] Preparations for verification of syntax tree on SILGen tests

### PR DESCRIPTION
**Pull request by ahoppen**
These are a couple of minor fixes that will allow libSyntax to parse all the SILGen tests, in particular:
- Add a `-verify-syntax-tree` frontend flag that will parse the syntax tree and error if any unknown nodes exist within it
- libSyntax fixes
  - Allow key paths on specialised generics (e.g. `Array<Int>.[]`)
  - Allow computed properties with non-standard accessors (e.g. `unsafeMutableAddress`)
  - Fix the parsing of delayed function bodies (I think this would never surface to the user but came up with the SILGen tests so I fixed it)
- SILGen test updates
  - Update operator declaration to Swift 4 syntax (without `{}`)
  - Update generic where clauses to Swift 4 syntax (trailing after the function signature instead of in the generics list)
  - Remove a superflous `)` in one test

In an upcoming PR I will flip the switch to always verify the syntax tree generated when parsing the SILGen tests. That way we should have a smoke test making sure that no new syntax gets introduced without libSyntax being able to parse it.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@eeckstein could you take a look at the update of sil gen tests? @ahoppen makes all silgen tests also libSyntax tests by ensuring we have no problem parsing those valid syntax.

**Review by eeckstein — APPROVED** [review]
lgtm.
@jckarter: any comments from your side?

File: utils/gyb_syntax_support/DeclNodes.py [hunk 1]
```diff
@@ -476,7 +476,12 @@
              Child('Modifier', kind='DeclModifier', is_optional=True),
              Child('AccessorKind', kind='Token',
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply utils/gyb_syntax_support/DeclNodes.py]: I'll do that in a follow-up PR.

File: lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -3870,7 +3870,12 @@ static ParameterList *parseOptionalAccessorArgument(SourceLoc SpecifierLoc,
   return ParameterList::create(P.Context, StartLoc, param, EndLoc);
 }
 
-static unsigned skipUntilMatchingRBrace(Parser &P) {
+static unsigned skipUntilMatchingRBrace(Parser &P,
+                                        SyntaxParsingContext *&SyntaxContext) {
+  SyntaxParsingContext BlockItemListContext(SyntaxContext,
+                                            SyntaxKind::CodeBlockItemList);
+  SyntaxParsingContext BodyContext(SyntaxContext, SyntaxKind::CodeBlock);
+  BodyContext.setDiscard();
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/Parse/ParseDecl.cpp]: Good point. I thought it was only used in delayed function body parsing. Changed it to a `TokenList`.

File: test/SILGen/enum_generic_raw_value.swift [hunk 3]
```diff
@@ -6,6 +6,6 @@ enum E<T>: Int {
 }
 
 // CHECK-LABEL: sil hidden @$S22enum_generic_raw_value1FO
-enum F<T: ExpressibleByIntegerLiteral where T: Equatable>: T {
+enum F<T: ExpressibleByIntegerLiteral>: T where T: Equatable {
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply test/SILGen/enum_generic_raw_value.swift]: I spoke with Argyrios about this and we decided not to support libSyntax parsing of Swift 3. It is being deprecated anyway and we will probably not be able to provide the coverage we will for Swift 4.

File: test/SILGen/protocol_resilience.swift [hunk 4]
```diff
@@ -5,10 +5,10 @@
...
 import resilient_protocol
 
-prefix operator ~~~ {}
-infix operator <*> {}
-infix operator <**> {}
-infix operator <===> {}
+prefix operator ~~~
+infix operator <*>
+infix operator <**>
+infix operator <===>
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply test/SILGen/protocol_resilience.swift]: What do you mean?

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: test/SILGen/protocol_resilience.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply test/SILGen/protocol_resilience.swift]: I wanted to accept `where` clause in generic parameter clause, but I do *not* wanted to accept deprecated `operator` decl syntax.

**Review by rintaro — APPROVED** [review]

**Comment by jckarter** [comment]
@eeckstein SILGen test changes look fine to me.

**Pull request merged**